alf/utils/gin_utils.py

```diff
@@ -53,3 +56,93 @@ def inoperative_config_str(max_line_length=80, continuation_indent=4):
     gin.config._OPERATIVE_CONFIG = operative_config
     gin.config._IMPORTED_MODULES = imported_module
     return inoperative_str
+
+
+def _config_gin_eval(func):
+    """Decorator to make function `gin_eval` configurable.
+    And it get and pass expected frame context variables `globals` and `locals`
+    to `func` when is actually called. See `gin_eval` for more details.
+    """
+
+    name = func.__name__
+    module = getattr(func, '__module__', None)
+    selector = module + '.' + name if module else name
+    fn = gin.config._ensure_wrappability(func)
+
+    @six.wraps(fn)
+    def gin_wrapper(**kwargs):
+        scope_components = gin.config.current_scope()
+        new_kwargs = {}
+        for i in range(len(scope_components) + 1):
+            partial_scope_str = '/'.join(scope_components[:i])
+            new_kwargs.update(
+                gin.config._CONFIG.get((partial_scope_str, selector), {}))
+        scope_str = partial_scope_str
+        operative_parameter_values = {}
+        operative_parameter_values.update(new_kwargs)
+        gin.config._OPERATIVE_CONFIG.setdefault(
+            (scope_str, selector), {}).update(operative_parameter_values)
+        new_kwargs = copy.deepcopy(new_kwargs)
+        # hack here ,injection was actually called at frame_trace[-7]
+        # (gin-config>=0.1.3,<=0.3.3)
+        frame = sys._getframe(7)
```

Comment:
7 seems very dangerous here. Is there any way to prevent error?

Reply:
Yes, only for gin-config>=0.1.3,<=0.3.3.  I'll  try to find a better way to extend gin config